modules/core/src/test/java/org/apache/ignite/cdc/WalForCdcTest.java

```diff
@@ -0,0 +1,250 @@
+import static org.apache.ignite.testframework.GridTestUtils.waitForCondition;
+
+/** Check only {@link DataRecord} written to the WAL for in-memory cache. */
+@RunWith(Parameterized.class)
+public class WalForCdcTest extends GridCommonAbstractTest {
+    /** */
+    private static final int RECORD_COUNT = 10;
+
+    /** */
+    @Parameterized.Parameter
+    public CacheMode mode;
+
+    /** */
+    @Parameterized.Parameter(1)
+    public CacheAtomicityMode atomicityMode;
+
+    /** */
+    private boolean persistenceEnabled;
+
+    /** */
+    private boolean cdcEnabled;
+
+    /** */
+    @Parameterized.Parameters(name = "mode={0}, atomicityMode={1}")
+    public static Collection<?> parameters() {
+        List<Object[]> params = new ArrayList<>();
+
+        for (CacheMode mode : Arrays.asList(REPLICATED, PARTITIONED))
+            for (CacheAtomicityMode atomicityMode : Arrays.asList(ATOMIC, TRANSACTIONAL))
+                params.add(new Object[] {mode, atomicityMode});
+
+        return params;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setDataStorageConfiguration(new DataStorageConfiguration()
+            .setWalMode(WALMode.FSYNC)
+            .setWalForceArchiveTimeout(WAL_ARCHIVE_TIMEOUT)
+            .setDefaultDataRegionConfiguration(new DataRegionConfiguration()
+                .setPersistenceEnabled(persistenceEnabled)
+                .setCdcEnabled(cdcEnabled)));
+
+        cfg.setConsistentId(igniteInstanceName);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        stopAllGrids();
+
+        cleanPersistenceDir();
+
+        cdcEnabled = true;
+        persistenceEnabled = false;
+    }
+
+    /** */
+    @Test
+    public void testOnlyDataRecordWritten() throws Exception {
+        IgniteEx ignite = startGrid(0);
+
+        ignite.cluster().state(ClusterState.ACTIVE);
+
+        AtomicInteger cntr = new AtomicInteger();
+
+        // Check only `DataRecords` written in WAL for in-memory cache with CDC enabled.
+        doTestWal(ignite, cache -> {
+            for (int i = 0; i < RECORD_COUNT; i++)
+                cache.put(keyForNode(ignite.affinity(DEFAULT_CACHE_NAME), cntr, ignite.localNode()), i);
+        });
+
+        // Check no WAL written during rebalance.
+        IgniteEx ignite1 = startGrid(1);
+
+        awaitPartitionMapExchange(false, true, null);
+
+        // Can't use `waitForCondition` because if test passed
+        // then no `DataRecords` loged therefore no segment archivation.
+        Thread.sleep(3 * WAL_ARCHIVE_TIMEOUT);
+
+        int walRecCnt = checkDataRecords(ignite1);
+
+        assertEquals(0, walRecCnt);
+
+        // Check `DataRecords` written on second node after rebalance.
+        doTestWal(ignite1, cache -> {
```

Comment:
Let's add a check for ignite0 too, we will test that rebalance doesn't break CDC on existing nodes.

Reply:
Good point. Check added.